Adds WitnessPersistence

gtauzin opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/giotto-learn/giotto-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### What does this implement/fix? Explain your changes.
Adds WitnessPersistence from GUDHI's bindings.

Commit:
Add WitnessPersistence

Commit:
[WIP] Witness awaiting for landmarks methods

Commit:
[WIP] Improve tests for WitnessPersistence

Commit:
Add bindings for Subsampling of Gudhi

there's a current discussion on how we resolve CGAL and EIGEN dependency

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Add python Interface for Subsampling

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Add test for subsampling python interface

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Merged with upstream

Commit:
Merge pull request #3 from reds-heig/add_subsampling

Add subsampling

Commit:
Merge with upstream master

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Add WitnessPersistence to doc index

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Add WitnessPersistence

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Add subsamplings file

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Fix for tests

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Merge branch 'master' of github.com:giotto-ai/giotto-tda into witness

ulupo committed:
Linting

Commit:
Fix errors

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Merge branch 'witness' of github.com:gtauzin/giotto-learn-1 into witness

gtauzin committed:
Add create_nearest_landmark_table function

gtauzin committed:
Fix tests

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Fix linting

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge branch 'master' of github.com:giotto-ai/giotto-tda into witness

gtauzin committed:
Ignore subsampling tests and test output

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Skip subsampling tests instead of ignoring them

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge with upstream master

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo commented on gtda/homology/simplicial.py:
```diff
@@ -684,6 +685,7 @@ def transform(self, X, y=None):
                                    self.infinity_values_, self.n_jobs)
         return Xt
 
+<<<<<<< HEAD
```
This looks dangerous...

gtauzin committed:
Fix merging

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge branch 'master' of github.com:giotto-ai/giotto-tda into witness

gtauzin committed:
Fix wrong import

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge branch 'master' of github.com:giotto-ai/giotto-tda into witness

gtauzin committed:
Merge hera submodule

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge ripser from upstream master

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

MonkeyBreaker commented on CMakeLists.txt:
```diff
@@ -307,3 +307,29 @@ else()
+#   target_link_libraries(gtda_subsampling PRIVATE)
+# endif()
+
+# target_compile_definitions(gtda_subsampling PRIVATE BOOST_RESULT_OF_USE_DECLTYPE=1 BOOST_ALL_NO_LIB=1 BOOST_SYSTEM_NO_DEPRECATED=1)
```
Why add a module commented ?
We should only push code uncommented or with a comment explaining why it's commented.
Otherwise, in the future we'll be afraid of removing it because we don't know why it's commented ...

gtauzin replied:
I left all the functions requiring CGAL that you merged on the branch. Indeed, I agree that all comments should be removed. Should the subsampling functions you wrote be removed too? My understanding was that CGAL was going to be a dependency at one point.

MonkeyBreaker commented on gtda/externals/python/__init__.py:
```diff
@@ -6,3 +6,5 @@
 from .strong_witness_complex_interface import StrongWitnessComplex
 from .rips_complex_interface import RipsComplex, SparseRipsComplex
 from .cech_complex_interface import CechComplex
+# from .subsampling_interface import choose_n_farthest_points,\
+#     pick_n_random_points, sparsify_point_set
```
Same remark as with CMakeLists.txt

MonkeyBreaker commented on gtda/externals/bindings/subsampling_bindings.cpp:
```diff
@@ -0,0 +1,46 @@
+            &Gudhi::subsampling::subsampling_sparsify_points));
+  m.def("subsampling_sparsify_points_from_file",
+        py::overload_cast<const std::string&, double>(
+            &Gudhi::subsampling::subsampling_sparsify_points_from_file));
```
Now that I have more experience with `pybind11`, I try to add two more information to the `m.def(...)`:

* How to give [arguments names](https://pybind11.readthedocs.io/en/stable/basics.html#keyword-arguments), I have a preference for the shorthand notation `"<variable_name>"_a`, you can also specify default values if it's necessary
* Hot to [document functions](https://pybind11.readthedocs.io/en/stable/basics.html#creating-bindings-for-a-simple-function), last argument of `m.def` can be a short description of the function, add it if the name of the function is not clear enough.

gtauzin replied:
Please go ahead, you can make a PR to my fork's branch

MonkeyBreaker replied:
Done, I pushed on your branch directly if you don't mind

MonkeyBreaker commented on gtda/externals/python/tests/test_subsampling.py:
```diff
@@ -0,0 +1,109 @@
+                    found = True
+            # Check each sub set point is existing in the point set
+            assert found is True
+
+
+@pytest.mark.skip(reason="requires CGAL")
+def test_simple_pick_n_random_points():
+    point_set = [[0, 1], [0, 0], [1, 0], [1, 1]]
+    # Test the limits
+    assert pick_n_random_points(points=[], nb_points=0) == []
+    assert pick_n_random_points(points=[], nb_points=1) == []
+    assert pick_n_random_points(points=point_set, nb_points=0) == []
+
+    # Go furter than point set on purpose
+    for iter in range(1, 10):
+        sub_set = pick_n_random_points(points=point_set, nb_points=iter)
+        print(5)
+        for sub in sub_set:
+            found = False
+            for point in point_set:
+                if point == sub:
+                    found = True
+            # Check each sub set point is existing in the point set
+            assert found is True
+
+
+@pytest.mark.skip(reason="requires CGAL")
+def test_simple_sparsify_points():
+    point_set = [[0, 1], [0, 0], [1, 0], [1, 1]]
+    # Test the limits
+    assert sparsify_point_set(points=point_set, min_squared_dist=0.0) ==\
+        point_set
+    assert sparsify_point_set(points=point_set, min_squared_dist=1.0) ==\
+        point_set
+    assert sparsify_point_set(points=point_set, min_squared_dist=2.0) == [
+        [0, 1],
+        [1, 0],
+    ]
+    assert sparsify_point_set(points=point_set, min_squared_dist=2.01) ==\
+        [[0, 1]]
```
I am confused if the test of a module currently disable should be pushed.
Because currently this test are not run, and when the situation will change in regard to CGAL, maybe those test will be deprecated.
@ulupo What do you think about this ?

MonkeyBreaker commented on gtda/externals/python/subsampling_interface.py:
```diff
@@ -0,0 +1,110 @@
+            return subsampling_n_random_points_from_file(off_file.encode('utf-8'),
+                nb_points)
+        else:
+            print("file " + off_file + " not found.")
+    else:
+        if points is None:
+            # Empty points
+            points = []
+        return subsampling_n_random_points(points, nb_points)
+
+
+def sparsify_point_set(points=None, off_file='', min_squared_dist=0.0):
+    """Outputs a subset of the input points so that the squared distance
+    between any two points is greater than or equal to min_squared_dist.
+
+    :param points: The input point set.
+    :type points: vector[vector[double]].
+
+    Or
+
+    :param off_file: An OFF file style name.
+    :type off_file: string
+
+    :param min_squared_dist: Minimum squared distance separating the output \
+    points.
+    :type min_squared_dist: float.
+    :returns:  The subsample point set.
+    :rtype: vector[vector[double]]
+    """
+    if off_file:
+        if os.path.isfile(off_file):
+            return subsampling_sparsify_points_from_file(off_file.encode('utf-8'),
+                                                         min_squared_dist)
+        else:
+            print("file " + off_file + " not found.")
+    else:
+        if points is None:
+            # Empty points
+            points = []
+        return subsampling_sparsify_points(points, min_squared_dist)
```
I would add a header to the file explaining why this is unused in giotto-tda.
I'm not even sure dead code should be pushed

gtauzin replied:
If those are the bindings for a Transformer to appear in a future release, I think it is fine. But I am not sure what is the situation with CGAL right now?

MonkeyBreaker commented on CMakeLists.txt:
```diff
@@ -307,3 +307,29 @@ else()
     target_compile_options(gtda_cech_complex PUBLIC $<$<CONFIG:DEBUG>:-O2 -ggdb -D_GLIBCXX_DEBUG>)
 endif()
 
+target_compile_definitions(gtda_cech_complex PRIVATE BOOST_RESULT_OF_USE_DECLTYPE=1 BOOST_ALL_NO_LIB=1 BOOST_SYSTEM_NO_DEPRECATED=1)
```
Why it was necessary to add this `target_compile_definitions`, it's already in original file

gtauzin replied:
I will fix it, I believe it is  mistake that happened during merging!

MonkeyBreaker commented on .gitignore:
```diff
@@ -41,6 +41,7 @@ gtda/externals/pybind11
 # Unit test
 .pytest_cache/
 .hypothesis/
+test-output.xml
```
Nice !

Commit:
Add argument identifiers and quick documentation of bindings

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Update Subsampling cmake rules and disable them by default

Future work will decide if we keep it or not, but at leats code is not commented

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

ulupo committed:
Merge branch 'master' into witness

ulupo committed:
Merge branch 'master' into witness
